juju/juju
# [JUJU-4421] Fix issue getting secret watchers from facade context

Fix secret watcher breakage introduced by https://github.com/juju/juju/pull/15817

Plus a few drive by fixes described below.

The core issue is that in the Next() facade calls, we were fetching the secret watchers like this

`context.WatcherRegistry().Get(context.ID())`

instead of

`GetWatcherByID(context.WatcherRegistry(), context.Resources(), context.ID())`

The result was that Next() would error out with a not found error and workers which used secret watchers would break.

`GetWatcherByID` is deprecated, but for now is has been used to fix this issue since all the other watchers use it too and they can be updated together later.

The offer status watcher is also tweaked to use `EnsureErr()` when needed, rather than copying and pasting the EnsureErr code. EnsureErr is used all over - if we want to change that, we should do it in one go to avoid ad hoc code duplication everywhere.

The watcher issue affected cross model relations. As part of investigating the issue, a few small cleanups were done. The `RemoteRelationChangeEvent` strict gets `ApplicationToken` renamed to `ApplicationOrOfferToken` because it now can be either.
The `AppNameForOffer()` method is removed as we always process a remote app, not an offer.

And the event publish code in the remoteRelationsWorker is tweaked to match that in the remoteUnitsWorker for consistency.


## QA steps

deploy juju-qa-dummy-source
set the token
make an offer
make a new model
deploy juju-qa-dummy-sink
relate to the offer
ensure everything goes active
